mllib/src/main/scala/org/apache/spark/ml/feature/PowerTransform.scala

```diff
@@ -0,0 +1,561 @@
+      val localNumBins = $(numBins)
+      pairCounts
+        .groupBy("col")
+        .count()
+        .as[(Int, Long)]
+        .flatMap { case (col, num) =>
+          val group = num / localNumBins
+          if (group >= 2) {
+            Some((col, group))
+          } else {
+            None
+          }
+        }.collect().toMap
+    } else Map.empty[Int, Long]
+
+    if (groups.nonEmpty) {
+      pairCounts = makeBins(pairCounts.as[(Int, Double, Long)], groups)
+        .toDF("col", "value", "cnt")
+    }
+
+    val solutions = pairCounts
+      .groupBy("col")
+      .agg(collect_list(struct("value", "cnt")))
+      .as[(Int, Seq[(Double, Long)])]
+      .map { case (col, seq) =>
+        val nnz = seq.iterator.map(_._2).sum
+        val nz = numRows - nnz
+        val (solution, _) = localModelType match {
+          case BoxCox =>
+            require(nz >= 0)
+            val computeIter = if (nz > 0) {
+              () => seq.iterator ++ Iterator.single((0.0, nz))
+            } else {
+              () => seq.iterator
+            }
+            solveBoxCox(computeIter)
+          case YeoJohnson =>
+            require(nz == 0)
+            val computeIter = () => seq.iterator
+            solveYeoJohnson(computeIter)
+        }
+        (col, solution)
+      }.collect().toMap
+
+    val lambda = Array.ofDim[Double](numFeatures)
+    solutions.foreach { case (col, solution) => lambda(col) = solution }
+
+    if (solutions.size < numFeatures) {
+      localModelType match {
+        case YeoJohnson =>
+          // if some column only contains 0 values in YeoJohnson
+          val computeIter = () => Iterator.single((0.0, numRows))
+          val (zeroSolution, _) = solveYeoJohnson(computeIter)
+          Iterator.range(0, numFeatures)
+            .filterNot(solutions.contains)
+            .foreach { col => lambda(col) = zeroSolution }
+
+        case BoxCox =>
+          // This should never happen.
+          throw new IllegalArgumentException("BoxCox requires positive values")
+      }
+    }
+
+   copyValues(new PowerTransformModel(uid, Vectors.dense(lambda).compressed)
+    .setParent(this))
+  }
+
+  override def copy(extra: ParamMap): PowerTransform = defaultCopy(extra)
+
+  override def transformSchema(schema: StructType): StructType = {
+    validateAndTransformSchema(schema)
+  }
+}
+
+
+@Since("3.1.0")
+object PowerTransform extends DefaultParamsReadable[PowerTransform] {
+
+  override def load(path: String): PowerTransform = super.load(path)
+
+  /** String name for Box-Cox transform model type. */
+  private[feature] val BoxCox: String = "box-cox"
+
+  /** String name for Yeo-Johnson transform model type. */
+  private[feature] val YeoJohnson: String = "yeo-johnson"
+
+  /* Set of modelTypes that PowerTransform supports */
+  private[feature] val supportedModelTypes = Array(BoxCox, YeoJohnson)
+
+  private[feature] def brentSolve(obj: UnivariateFunction): (Double, Double) = {
```

Comment:
vs scikit-learn's [implementation](https://github.com/scikit-learn/scikit-learn/blob/b189bf60708af22dde82a00aca7b5a54290b666d/sklearn/preprocessing/_data.py#L3042):
use same `BrentAbs = 1.48E-8`;
sklearn uses bound [-2, 2], but it said "Providing the pair (xa,xb) does not always mean
        the obtained solution will satisfy xa<=x<=xb.";
sklearn use iters=500;